File: algorithms/2_medium/F/03524_find_x_value_of_array_i/dp.c
#define K_MAX 5
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
long long* resultArray(const int *restrict nums, const int len, int k, int *return_len) {
    assert(k >= 1 && k <= K_MAX && return_len != NULL);
    long long *restrict res = malloc((size_t)k * sizeof(*res));
    if (res == NULL) {
        *return_len = -1;
        return NULL;
    }
    long long prev[K_MAX] = { 0 };
    long long sum[K_MAX] = { 0 };
    for (int i = 0; i < len; i += 1) {
        long long curr[K_MAX] = { 0 };
        const int rem = nums[i] % k;
        curr[rem] += 1;
        for (int j = 0; j < k; j += 1) {
            curr[(j * rem) % k] += prev[j];
        }
        for (int j = 0; j < k; j += 1) {
            sum[j] += curr[j];
            prev[j] = curr[j];
        }
    }

    for (int i = 0; i < k; i += 1) {
        res[i] = sum[i];
    }
    *return_len = k;
    return res;
}

File: algorithms/2_medium/F/03524_find_x_value_of_array_i/dp.rs
impl Solution {
    const K_MAX: usize = 5;

    pub fn result_array(nums: Vec<i32>, k: i32) -> Vec<i64> {
        debug_assert!(k >= 1 && (k as usize) <= Self::K_MAX);
        let len = nums.len();
        let mut prev = [0_i64; Self::K_MAX];
        let mut sum = [0_i64; Self::K_MAX];
        for num in nums {
            debug_assert!(num >= 1);
            let mut curr = [0_i64; Self::K_MAX];
            let rem = num % k;
            curr[rem as usize] += 1;
            for i in 0..k {
                curr[((i * rem) % k) as usize] += prev[i as usize];
            }
            for i in 0..k {
                sum[i as usize] += curr[i as usize];
            }
            prev = curr;
        }

        sum[0..(k as usize)].to_vec()
    }
}
